src/adapter/src/coord.rs

```diff
@@ -2104,6 +2109,70 @@ impl Coordinator {
             ),
         );
 
+        debug!("coordinator init: initializing migrated builtin tables");
+        // When 0dt is enabled, we create new shards for any migrated builtin storage collections.
+        // In read-only mode, the migrated builtin tables (which are a subset of migrated builtin
+        // storage collections) need to be back-filled so that any dependent dataflow can be
+        // hydrated. Additionally, these shards are not registered with the txn-shard, and cannot
+        // be registered while in read-only, so they are written to directly.
+        let migrated_updates_fut = if self.controller.read_only() {
+            let min_timestamp = Timestamp::minimum();
+            let migrated_builtin_table_updates: Vec<_> = builtin_table_updates
+                .drain_filter_swapping(|update| {
+                    migrated_storage_collections_0dt.contains(&update.id)
+                        && self
+                            .controller
+                            .storage_collections
+                            .collection_frontiers(update.id)
+                            .expect("all tables are registered")
+                            .write_frontier
+                            .elements()
+                            == &[min_timestamp]
+                })
+                .collect();
+            if migrated_builtin_table_updates.is_empty() {
+                futures::future::ready(()).boxed()
+            } else {
+                let mut appends: BTreeMap<GlobalId, Vec<(Row, Diff)>> = BTreeMap::new();
+                for update in migrated_builtin_table_updates {
+                    appends
+                        .entry(update.id)
+                        .or_default()
+                        .push((update.row, update.diff));
+                }
+                for (_, updates) in &mut appends {
+                    differential_dataflow::consolidation::consolidate(updates);
+                }
+                info!(
+                    "coordinator init: rehydrating migrated builtin tables in read-only mode: {:?}",
+                    appends.keys().collect::<Vec<_>>()
+                );
+                let appends = appends
+                    .into_iter()
+                    .map(|(id, updates)| {
+                        let updates = updates
+                            .into_iter()
+                            .map(|(row, diff)| TimestamplessUpdate { row, diff })
+                            .collect();
+                        (id, updates)
+                    })
+                    .collect();
+                let fut = self
+                    .controller
+                    .storage
+                    .append_table(min_timestamp, boot_ts.step_forward(), appends)
```

Comment:
Why is it safe/correct to call append_table if the controller is in read only? Is this read only controller guaranteed to generate unique shard ids for its migrations, and so it only writes to data that it has exclusive knowledge of?

Reply:
> Is this read only controller guaranteed to generate unique shard ids for its migrations, and so it only writes to data that it has exclusive knowledge of?

That's the idea yes, but the whole thing is a bit hacky. We have a check here:

https://github.com/MaterializeInc/materialize/pull/28434/files#diff-3d40535ccc11f976645c3c15ed4b6b1a08b7c04de0d21085ba2a4a891b6e5422R1489-R1497

That's meant to throw an error if the shard isn't one of the newly migrated shards. Only newly migrated shards don't have a txn shard set which is what the conditional is checking.